internal/wrr/edf.go

```diff
@@ -87,6 +93,8 @@ func (edf *edfWrr) Next() interface{} {
 	}
 	item := edf.items[0]
 	item.deadline = edf.currentTime() + 1.0/float64(item.weight)
+	item.orderOffset = edf.currentOrderOffset
```

Comment:
I thought we wouldn't need to add another field just for this.
Instead of updating the value of the item (line 94, `item := edf.items[0]`) and then heap.Fix, if we pop the item, update it, and insert back, the order would be fixed.

Can you give it a try and see if it works? Thanks.

Reply:
This seems not work for the test with 3 elements
```
        wrr := NewEDF()
	wrr.Add("1", 1)
	wrr.Add("2", 1)
	wrr.Add("3", 1)
	expected := []string{"1", "2", "3", "1", "2", "3", "1", "2", "3", "1", "2", "3"}
	for i := 0; i < len(expected); i++ {
		item := wrr.Next().(string)
		t.Logf("wrr Next=%s, want=%s", item, expected[i])
		if item != expected[i] {

		}
	}
```
We got 131212323131

At the very beginning, the priority_queue of (item, deadline) looks like: {("1", 1), ("2", 2), ("3", 2)} After first Next, it will Pop item 1 and re-push it with deadline 2. Second Next, pop item 3, re-push with deadline 3. 

The init queue is: {("1", 1), ("2", 2), ("3", 2)} is because current implementation looks at the first item as current time. and golang Heap seems not maintain the FIFO itself. If we'd like to keep deadline strictly increasing, we need extra information.

Reply:
Looking at the result for 3 items (or, more than 2 items), I'm even less convinced this fix gets any advantages other than that the result looks betters.

If the fix is to just move code around, or to call a different method, I would be OK.
But if to get this, we need to add a new field, and it doesn't really gain any features in terms of functionality, I would prefer not doing it. In the end, the result is WRR either way.

Reply:
I agree with you that without this fix it's still correct WRR. I am worried in one case, when there are huge amount of clients but very low qps per client, some server/endpoint may get more load compare to simple RR. For the 3 elements example, "13121" (current wrr impl) versus "12312". this issue can also be addressed by shuffle subchannel list though.

Reply:
to add another point is, degenerate to simple RR is easier to reason (imho), and to match the envoy behavior https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/edf_scheduler.h#L80 . 

Is there concerns about adding a new field?

Reply:
Re imbalance between backends with low qps clients: this change doesn't make this situation better. It is a different problem to solve.
And with or without this change, you would get backends in groups (`1{312}{213}{123}`, order might be wrong, but you get the idea, only the first pick is "problematic"), so the imbalance problem you mentioned with `13121` won't be a big deal later on.

Re concerns about adding a new field: not really a big concern. Just that if this new field doesn't do much, adding it just makes code harder to read.

Re the offset fields in envoy: is this offset ever mentioned any spec?
But since we are basically trying to mimic envoy behavior, adding a field similar to that SG.
Though it seems the offset field added in this PR behaves quite different from the one in envoy. Can you double check?

Reply:
Fair, in the long run, it's fine.

My understand of the `_order_offset` in envoy code is to maintain FIFO order, and it's stated in the comment one line above. should be the same as this PR, no?

yishu

Reply:
It seems `offset` doesn't need to be changed in `Next`. Only `Push` to increment it, does that sound right?

Reply:
ah, this wrr is not exactly the same as edf scheduler in envoy

envoy does another push after pick https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/load_balancer_impl.cc#L751-L752
and https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/load_balancer_impl.cc#L794-L796

Do we expect to do exactly the same as envoy?

Reply:
Envoy's pick removes the entry, so they need to push it back. In this EDF, we do a `heap.Fix` instead.
It seems envoy also support updating weights, that's currently a missing feature in this EDF.

Test failure, I think, is because of how we get `currentTime()` during `Add`, and that messes up the first round of picks?
Adding a member for current time (thus making it more similar to envoy's behavior) would be good.

Reply:
yeah, that's exactly what I am asking. we need an extra field. 

> It seems envoy also support updating weights, that's currently a missing feature in this EDF.
is there an issue for this?  We need this feature for our usage.

Reply:
There's no issue. Feel free to file one.

And when you need this feature, how are you planning to use this edf. It's an internal package.

Reply:
ah sorry, I think my previous comment is not accurate. It's more like "update weight" is useful for eDS support from our experience.